Implement nested markdown lists

MichaelHatherly opened the pull request:
Adds basic support for writing nested lists, as well as any other "block"-level markdown element such as code blocks, tables, quote blocks, etc. within a single list item.

Since lists can now contain block elements some additional spacing has been added between items when output in the terminal to make it easier to differentiate between elements.

Fixes #16003.

---

`genstdlib` changes appear to all be simple whitespace changes that don't affect the final docs. ~~(Only exception I've found is the `IOContext` change where an indented code block is lost due to a list appearing before it. This can just be replaced with a fenced code block though, so not a problem really.)~~

~~Still todo: enable starting ordered lists from numbers other than `1`, which just needs several changes throughout the `markdown/render` folder.~~

MichaelHatherly committed:
Implement nested markdown lists

Adds basic support for writing nested lists, as well as any
other "block"-level markdown element such as code blocks,
tables, quote blocks, etc. within a single list item.

Since lists can now contain block elements some additional spacing
has been added between items when output in the terminal to make
it easier to differentiate between elements.

MichaelHatherly commented:
And #11249 could be closed with this change:

```diff
diff --git a/base/markdown/Common/block.jl b/base/markdown/Common/block.jl
index da53983..386207d 100644
--- a/base/markdown/Common/block.jl
+++ b/base/markdown/Common/block.jl
@@ -226,6 +226,7 @@ List(xs...) = List(vcat(xs...))
 const BULLETS = r"^ {0,3}(\*|\+|•|-)( |$)"
 const NUM_OR_BULLETS = r"^ {0,3}(\*|•|\+|-|\d+(\.|\)))( |$)"
 
+@breaking true ->
 function list(stream::IO, block::MD)
     withstream(stream) do
         bullet = startswith(stream, NUM_OR_BULLETS; eat = false)
```

If there's any support for that then I'm happy to add it as well.

tkelman commented on doc/stdlib/io-network.rst:
```diff
@@ -415,9 +415,7 @@ General I/O
-   .. code-block:: julia
-
-       IOContext(io::IO, context::IOContext)
+     IOContext(io::IO, context::IOContext)
```
maybe these should be separate docstrings?

MichaelHatherly replied:
Yes, was wondering whether that would be better perhaps. Should I split them in this PR or just fix up the code block for now?

tkelman replied:
either would work, separating the docstrings could technically be a separate PR to be pedantic

tkelman commented on doc/stdlib/parallel.rst:
```diff
@@ -173,29 +173,19 @@ General Parallel Computing Support
-
-   * ``exename``\ : name of the ``julia`` executable. Defaults to ``"$JULIA_HOME/julia"`` or              ``"$JULIA_HOME/julia-debug"`` as the case may be.
-
+     sshflags=``-i /home/foo/bar.pem``
```
don't think this is right, should be keeping the code for the whole line and the single-backticks as julia code if possible

MichaelHatherly replied:
Same problem as with `IOContext`, using a fenced code block should fix this too.

MichaelHatherly committed:
Add arbitrary starting values for lists

This allows ordered lists to be numbered starting from any
non-negative integer such as:

    7. foo
    8. bar
    ...

MichaelHatherly committed:
Fix rendering of two code blocks

Indented code blocks appearing directly after a markdown list cannot
be handled correctly since the indentation of the list contents
takes precedence over the code block's indentation. To resolve this
we use a fenced code block instead.

MichaelHatherly committed:
Re-run genstdlib

MichaelHatherly commented:
Rebased, should be good to go now? This is probably the last markdown feature we needed to make the RST -> MD manual conversion as lossless as possible I think.

tkelman commented:
lgtm

MichaelHatherly commented:
win32 error looks unrelated, will merge shortly as is.
